qcore/asserts.py

```diff
@@ -353,6 +362,51 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         raise AssertionError(message)
 
 
+class AssertRaisesInstance(object):
+    """With-context that asserts that the code within the context raises a specified exception
+    instance."""
+
+    def __init__(self, *expected_exception_instances, **kwargs):
+        assert (
+            len(expected_exception_instances) >= 1
+        ), "You must specify an exception instance when using AssertRaisesInstance"
+        self.expected_exception_instances = expected_exception_instances
+        self.extra = kwargs.pop("extra", None)
+        assert_eq({}, kwargs)
+
+    def __enter__(self):
+        self.ctx = AssertRaises(
+            *[type(exception_instance) for exception_instance in self.expected_exception_instances],
+            extra=self.extra)
+        self.assert_raises = self.ctx.__enter__()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        ret = self.ctx.__exit__(exc_type, exc_val, exc_tb)
+        if ret:
+            expected_args = []
+            for exception_instance in self.expected_exception_instances:
+                if type(exception_instance) == exc_type:
+                    if exception_instance.args == self.assert_raises.expected_exception_found.args:
```

Comment:
You can use `exc_val` instead of `self.assert_raises.expected_exception_found`. Also, why not check `exception_instance == exc_val` directly?

Reply:
Ok updated. Equality comparison on the instances themselves would return False even if constructed identically, so we have to make do with checking that the type and args match instead.